File: MailProject.Infrastructure/Services/UserService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using MailProject.Application.Common.Interfaces;
using MailProject.Application.Common.Models;
using MailProject.Application.DTOs;
using MailProject.Infrastructure.Persistence;

namespace MailProject.Infrastructure.Services
{
    public class UserService : IUserService
    {
        private readonly MailDbContext _context;

        public UserService(MailDbContext context)
        {
            _context = context;
        }

        public async Task<CommonResponseMessage<List<UserDto>>> GetAllUsersAsync()
        {
            var users = await _context.Users
                .Include(u => u.Package)
                .OrderByDescending(u => u.CreatedAt) // Assuming BaseEntity has CreatedAt
                .Select(u => new UserDto
                {
                    Id = u.Id,
                    FullName = u.FullName,
                    Email = u.Email,
                    Role = u.Role.ToString(),
                    PackageName = u.Package != null ? u.Package.Name : "Paket Yok",
                    PackageEndDate = u.PackageEndDate,
                    IsVerified = u.IsVerified
                })
                .ToListAsync();

            return CommonResponseMessage<List<UserDto>>.Success(users);
        }

        public async Task<CommonResponseMessage<bool>> UpdateUserPackageAsync(Guid userId, Guid packageId, DateTime? packageEndDate = null)
        {
            var user = await _context.Users.FindAsync(userId);
            if (user == null) return CommonResponseMessage<bool>.Fail("Kullanıcı bulunamadı", 404);

            var package = await _context.Packages.FindAsync(packageId);
            if (package == null) return CommonResponseMessage<bool>.Fail("Paket bulunamadı", 404);

            user.PackageId = packageId;
            user.PackageStartDate = DateTime.UtcNow;
            
            if (packageEndDate.HasValue)
            {
                // Ensure the date is in UTC, if it's Unspecified, treat as Local then convert, or assume UTC.
                // Best practice: Frontend sends ISO string, binding parses it.
                // If it's pure Date (yyyy-MM-dd), it might be midnight.
                user.PackageEndDate = packageEndDate.Value.Kind == DateTimeKind.Utc ? packageEndDate.Value : packageEndDate.Value.ToUniversalTime();
            }
            else
            {
                user.PackageEndDate = DateTime.UtcNow.AddMonths(1);
            }

            await _context.SaveChangesAsync();

            return CommonResponseMessage<bool>.Success(true, "Kullanıcı paketi güncellendi.");
        }
    }
}

File: MailProject.Client/src/app/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface UserDto {
  id: string;
  fullName: string;
  email: string;
  role: string;
  packageName: string;
  packageEndDate: string;
  isVerified: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private apiUrl = `${environment.apiUrl}/users`;

  constructor(private http: HttpClient) { }

  getUsers(): Observable<any> {
    return this.http.get<any>(this.apiUrl);
  }

  updateUserPackage(userId: string, packageId: string, packageEndDate?: string): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}/${userId}/package`, { packageId, packageEndDate });
  }

  deleteUser(userId: string): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/${userId}`);
  }
}
